rangeserver/src/prefetching_buffer.rs

```diff
@@ -0,0 +1,946 @@
+use bytes::Bytes;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::{watch, Mutex};
+use uuid::Uuid;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum KeyState {
+    Fetched,        // Key has been fetched and is in the BTree
+    Loading(u64),   // Key has been requested and is being fetched
+    Requested(u64), // Key has been requested but fetching has not been started
+}
+
+#[derive(Default, Debug)]
+pub struct PrefetchingBuffer {
+    pub prefetch_store: Arc<Mutex<BTreeMap<Bytes, Bytes>>>, // stores key / value
+    pub key_state: Arc<Mutex<HashMap<Bytes, KeyState>>>,    // stores key -> current fetch state
+    pub transaction_keys: Arc<Mutex<HashMap<Uuid, BTreeSet<Bytes>>>>, // stores transaction_id -> set of requested keys
+    pub key_transactions: Arc<Mutex<HashMap<Bytes, BTreeSet<Uuid>>>>, // stores key -> set of transaction_ids
+    pub key_state_watcher: Arc<Mutex<HashMap<Bytes, watch::Receiver<KeyState>>>>, // key -> receiver for state changes
+    pub key_state_sender: Arc<Mutex<HashMap<Bytes, watch::Sender<KeyState>>>>, // key -> sender for state changes
+    pub fetch_sequence_number: Mutex<u64>,
```

Comment:
A mutex is probably fine, but consider using an atomic here:
https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.fetch_add